Allow theming of actions

rullzer opened the pull request:
Fixes #6773 eventually ;)

@owncloud/designers I need your help to do this properly. We will move from calculating the path in javascript to using the icons from the icon.css file. So that styling can be done properly.

However, I need your css FU to make it look nice.

rullzer committed:
Make file actions use icon CSS classes

This makes theming of file actions possible

rullzer commented:
Currently I'm styling the file actions which look like: 
![actions](https://cloud.githubusercontent.com/assets/45821/13106999/a22fca9a-d56b-11e5-825d-efee16c7b6d0.png)

I replaced some of the icons with the triangle so I know stuff is different.

@owncloud/designers could you take a look at the css there to make it all work again?

PVince81 commented on apps/files/js/fileactions.js:
```diff
@@ -11,7 +11,7 @@
 
 	var TEMPLATE_FILE_ACTION_TRIGGER =
-		'<a class="action action-{{nameLowerCase}}" href="#" data-action="{{name}}">' +
+		'<a class="action action-{{nameLowerCase}} icon-{{nameLowerCase}}" href="#" data-action="{{name}}">' +
```
Why not use "class" from the action object ?

PVince81 commented on apps/files/js/fileactions.js:
```diff
@@ -142,6 +142,7 @@
 				displayName: action.displayName,
 				mime: mime,
 				order: action.order || 0,
+				class: action.class,
```
use this instead of `nameLowerCase` in the template for the icon

PVince81 replied:
and here you could make the class name default to `"icon-" + action.name.toLowerCase()` if not set.
Maybe call it "iconClass" instead ?

PVince81 commented on apps/files/js/filelist.js:
```diff
@@ -2722,7 +2722,7 @@
 			var $newButton = $(this._addButtonTemplate({
 				addText: t('files', 'New'),
-				iconUrl: OC.imagePath('core', 'actions/add')
+    			class: 'icon-add',
```
indent fail

jancborchardt commented:
The example theme should only have the absolute necessary examples though to be as simple as possible. Action icons aren't part of that – and also the icons.css file should not be copied whole.

Is should be possible by simply adding different action items with the same name in the theme folder. If necessary, adding a CSS rule for those icons. But not duplicating the whole file.

rullzer commented:
@jancborchardt I wont merge it with the example stuff as is... but I aded them for developing so that we all see the same stuff.

rullzer commented:
@jancborchardt I do need your help here with the css stuff...

PVince81 commented:
@jancborchardt can you help @rullzer ?

PVince81 commented:
@rullzer it seems there is a problem with the logic.
The actions that have only "iconClass" but no "icon" are rendered as "no-icon" in the menu.
Please change it as follows:
- if neither "icon" not "iconClass" is specified, then render as "no-icon"
- make "iconClass" be empty if none specified, remove the default "icon-" + name logic.

PVince81 commented:
Ok, let me fix this myself, there's additional stuff needed.

PVince81 commented:
Adjusted and squashed.
The file actions icons, the "add" menu, the share icon replacing logic, have been adjusted to use "icon-X" CSS classes instead of "img" elements.

@rullzer I removed the example theme modifications, was confused.

I tried again with a local custom theme, and there is a problem with theming: for some reason, it refuses the load images from "themes/example/core/img/actions" (which I copied from core/img/actions).
I get a 403 Forbidden there. However loading images from "themes/example/core/img/"  works.

@LukasReschke is there a security restriction in place ?

PVince81 commented:
Never mind, my "img/actions" folder had the wrong permissions...

PVince81 commented:
@rullzer @nickvergessen @icewind1991 @schiesbn @jancborchardt @MorrisJobke please review.

To test:

1. `cp core/css/icons.css themes/example/core/css/`
1. `cp -Rv core/img/actions themes/example/core/img`
1. Check the permissions on the "themes/example/core/img/actions" (just in case, happened to me)
1. Maybe swap some icon images, for example `cp star.svg download.svg`
1. Open ownCloud
1. Open the file actions dropdown

You should see a star instead of the download icon.

PVince81 commented:
Also, some regression test needed for all share icon states:
- outgoing link share
- outgoing user share
- incoming share
- switching share state properly updates the share icon
- refresh the page, icons still there

I did test these and they worked.

rullzer commented:
Fixed the details icon.

rullzer commented:
- [ ] Bug: I can't see the 3 dots in the public link page (when sharing folders).

rullzer commented:
- [ ] Missing: Delete icon next to recipients of share is not themed yet

PVince81 commented:
I thought this was only about file actions...

rullzer commented:
I'm more than fine to leave the missing stuff for a future PR... this is already solving a lot now. Thanks for the fixes @PVince81 .

However the bug on the public link page we should fix

rullzer committed:
Close class

PVince81 commented:
Pfff... now looks like the icons don't have the correct size. The star icon is actually smaller than 16px but the CSS is making it look bigger. Will need more work...

PVince81 commented:
Or... it's the SVG rendering that looks different, weiiiird:

Star before: (img tag)
![star-before](https://cloud.githubusercontent.com/assets/277525/13325643/05cdf6ee-dbe3-11e5-9d3a-1b00f77419a9.png)

Star after: (icon CSS)
![star-after](https://cloud.githubusercontent.com/assets/277525/13325656/0e14c3b4-dbe3-11e5-9d6d-529a55a6282c.png)

Commit:
Make more action icons themable + style fixes

Action icons now appear properly in public link page.
More actions are now CSS icons.

PVince81 commented:
Ok, I had to add `background-size: 16px 16px`.

Please recheck @rullzer

PVince81 commented:
Will fix the JS unit test in a minute...

Commit:
Fix favorite icon unit test for icon image

PVince81 commented:
Done, please review!

rullzer commented:
:rocket: :+1: all looking good!

MorrisJobke commented:
Tested and works :+1:

MorrisJobke commented:
Rebased to re-trigger CI

jancborchardt commented:
Yep, looking good now!
